### ISSUE TITLE
Exiv2 and MinGW : The library compile fine with Unicode paths support

### ISSUE THREAD

**Issue by cgilles** [Exiv2/exiv2#468]
[gilles@localhost exiv2 (master)]$ ./bootstrap.mxe /mnt/data/GIT/6.x/project/bundles/mxe/build.win64
MXE target : 64 bits shared
Installing to /mnt/data/GIT/6.x/project/bundles/mxe/build.win64 for target x86_64-w64-mingw32.shared with build mode RelWithDebInfo and configure options 
== Using MXE wrapper: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-cmake
     - cmake version 3.10.0
     - warnings for unused CMAKE_POLICY_DEFAULT variables can be ignored
== Using MXE toolchain: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/share/cmake/mxe-conf.cmake
== Using MXE runresult: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/share/cmake/modules/TryRunResults.cmake
loading initial cache file /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/share/cmake/modules/TryRunResults.cmake
-- The CXX compiler identification is GNU 5.5.0
-- The C compiler identification is GNU 5.5.0
-- Check for working CXX compiler: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-g++
-- Check for working CXX compiler: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-gcc
-- Check for working C compiler: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Regex: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/include  
-- Found ZLIB: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/lib/libz.dll.a (found version "1.2.11") 
-- Found EXPAT: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/lib/libexpat.dll.a (found version "2.2.6") 
-- Found Intl: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/include  
-- Found Iconv: /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/lib/libiconv.dll.a  
-- ICONV_INCLUDE_DIR : /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/include
-- ICONV_LIBRARIES : /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared/lib/libiconv.dll.a
-- Looking for gmtime_r
-- Looking for gmtime_r - not found
-- Looking for mmap
-- Looking for mmap - not found
-- Looking for munmap
-- Looking for munmap - not found
-- Looking for strerror_r
-- Looking for strerror_r - not found
-- Looking for libintl.h
-- Looking for libintl.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for memory.h
-- Looking for memory.h - found
-- Looking for process.h
-- Looking for process.h - found
-- Looking for stdbool.h
-- Looking for stdbool.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for strings.h
-- Looking for strings.h - found
-- Looking for sys/mman.h
-- Looking for sys/mman.h - found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for regex.h
-- Looking for regex.h - found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
-- Install prefix:    /mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/x86_64-w64-mingw32.shared
-- ------------------------------------------------------------------
-- CMake Generator:   Unix Makefiles
-- Compiler info: GNU (/mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-g++) ; version: 5.5.0
--  --- Compiler flags --- 
-- General:            -Wall -Wcast-align -Wpointer-arith -Wformat-security -Wmissing-format-attribute -Woverloaded-virtual -W -std=gnu++98
-- Extra:              
-- Debug:             -g3 -gstrict-dwarf -O0
-- Release:           -O3 -DNDEBUG
-- RelWithDebInfo:    -O2 -g -DNDEBUG
-- MinSizeRel:        -Os -DNDEBUG
--  --- Linker flags --- 
-- General:           
-- Debug:             
-- Release:           
-- RelWithDebInfo:    
-- MinSizeRel:        
-- 
-- Compiler Options
-- Warnings as errors:                 NO
-- Use extra compiler warning flags:   NO
-- 
-- Compiler info: GNU (/mnt/data/GIT/6.x/project/bundles/mxe/build.win64/usr/bin/x86_64-w64-mingw32.shared-g++) ; version: 5.5.0
-- ------------------------------------------------------------------
-- Building shared library:            YES
-- Building PNG support:               YES
-- XMP metadata support:               YES
-- Native language support:            YES
-- Conversion of Windows XP tags:      YES
-- Nikon lens database:                YES
-- Building video support:             NO
-- Building webready support:          NO
-- Dynamic runtime override:           NO
-- Unicode paths (wstring):            YES
-- Building exiv2 command:             NO
-- Building samples:                   NO
-- Building PO files:                  NO
-- Building unit tests:                NO
-- Building doc:                       NO
-- Building with coverage flags:       NO
-- Using ccache:                       NO
-- ------------------------------------------------------------------
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/data/GIT/exiv2/build

[gilles@localhost exiv2 (master)]$ 
[gilles@localhost exiv2 (master)]$ cd build
[gilles@localhost build (master)]$ make
Scanning dependencies of target xmp
[  1%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/ExpatAdapter.cpp.obj
[  2%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/MD5.cpp.obj
[  3%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/ParseRDF.cpp.obj
[  4%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/UnicodeConversions.cpp.obj
[  6%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/WXMPIterator.cpp.obj
[  7%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/WXMPMeta.cpp.obj
[  8%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/WXMPUtils.cpp.obj
[  9%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XML_Node.cpp.obj
[ 10%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPCore_Impl.cpp.obj
[ 12%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPIterator.cpp.obj
[ 13%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPMeta-GetSet.cpp.obj
[ 14%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPMeta-Parse.cpp.obj
[ 15%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPMeta-Serialize.cpp.obj
[ 16%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPMeta.cpp.obj
[ 18%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPUtils-FileInfo.cpp.obj
[ 19%] Building CXX object xmpsdk/CMakeFiles/xmp.dir/src/XMPUtils.cpp.obj
[ 20%] Linking CXX static library ../lib/libxmp.a
[ 20%] Built target xmp
Scanning dependencies of target exiv2lib_int
[ 21%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/canonmn_int.cpp.obj
[ 22%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/casiomn_int.cpp.obj
[ 24%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/cr2header_int.cpp.obj
[ 25%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/crwimage_int.cpp.obj
[ 26%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/fujimn_int.cpp.obj
[ 27%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/makernote_int.cpp.obj
[ 28%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/minoltamn_int.cpp.obj
[ 30%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/nikonmn_int.cpp.obj
[ 31%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/olympusmn_int.cpp.obj
[ 32%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/orfimage_int.cpp.obj
[ 33%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/panasonicmn_int.cpp.obj
[ 34%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/pentaxmn_int.cpp.obj
[ 36%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/rw2image_int.cpp.obj
[ 37%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/samsungmn_int.cpp.obj
[ 38%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/sigmamn_int.cpp.obj
[ 39%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/sonymn_int.cpp.obj
[ 40%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/tags_int.cpp.obj
[ 42%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/tiffcomposite_int.cpp.obj
[ 43%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/tiffimage_int.cpp.obj
[ 44%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/tiffvisitor_int.cpp.obj
[ 45%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/helper_functions.cpp.obj
[ 46%] Building CXX object src/CMakeFiles/exiv2lib_int.dir/pngchunk_int.cpp.obj
[ 46%] Built target exiv2lib_int
Scanning dependencies of target exiv2lib
[ 48%] Building CXX object src/CMakeFiles/exiv2lib.dir/asfvideo.cpp.obj
[ 49%] Building CXX object src/CMakeFiles/exiv2lib.dir/basicio.cpp.obj
[ 50%] Building CXX object src/CMakeFiles/exiv2lib.dir/bigtiffimage.cpp.obj
[ 51%] Building CXX object src/CMakeFiles/exiv2lib.dir/bmpimage.cpp.obj
[ 53%] Building CXX object src/CMakeFiles/exiv2lib.dir/convert.cpp.obj
/mnt/data/GIT/exiv2/src/convert.cpp:1511:10: warning: 'bool {anonymous}::convertStringCharsetWindows(std::__cxx11::string&, const char*, const char*)' defined but not used [-Wunused-function]
     bool convertStringCharsetWindows(std::string& str, const char* from, const char* to)
          ^
[ 54%] Building CXX object src/CMakeFiles/exiv2lib.dir/cr2image.cpp.obj
[ 55%] Building CXX object src/CMakeFiles/exiv2lib.dir/crwimage.cpp.obj
[ 56%] Building CXX object src/CMakeFiles/exiv2lib.dir/datasets.cpp.obj
[ 57%] Building CXX object src/CMakeFiles/exiv2lib.dir/easyaccess.cpp.obj
[ 59%] Building CXX object src/CMakeFiles/exiv2lib.dir/epsimage.cpp.obj
[ 60%] Building CXX object src/CMakeFiles/exiv2lib.dir/error.cpp.obj
[ 61%] Building CXX object src/CMakeFiles/exiv2lib.dir/exif.cpp.obj
[ 62%] Building CXX object src/CMakeFiles/exiv2lib.dir/futils.cpp.obj
[ 63%] Building CXX object src/CMakeFiles/exiv2lib.dir/gifimage.cpp.obj
[ 65%] Building CXX object src/CMakeFiles/exiv2lib.dir/http.cpp.obj
[ 66%] Building CXX object src/CMakeFiles/exiv2lib.dir/image.cpp.obj
/mnt/data/GIT/exiv2/src/image.cpp:861:77: warning: unused parameter 'useCurl' [-Wunused-parameter]
     BasicIo::AutoPtr ImageFactory::createIo(const std::wstring& wpath, bool useCurl)
                                                                             ^
[ 67%] Building CXX object src/CMakeFiles/exiv2lib.dir/ini.cpp.obj
[ 68%] Building CXX object src/CMakeFiles/exiv2lib.dir/iptc.cpp.obj
[ 69%] Building CXX object src/CMakeFiles/exiv2lib.dir/jp2image.cpp.obj
[ 71%] Building CXX object src/CMakeFiles/exiv2lib.dir/jpgimage.cpp.obj
[ 72%] Building CXX object src/CMakeFiles/exiv2lib.dir/matroskavideo.cpp.obj
[ 73%] Building CXX object src/CMakeFiles/exiv2lib.dir/metadatum.cpp.obj
[ 74%] Building CXX object src/CMakeFiles/exiv2lib.dir/mrwimage.cpp.obj
[ 75%] Building CXX object src/CMakeFiles/exiv2lib.dir/orfimage.cpp.obj
[ 77%] Building CXX object src/CMakeFiles/exiv2lib.dir/pgfimage.cpp.obj
[ 78%] Building CXX object src/CMakeFiles/exiv2lib.dir/preview.cpp.obj
[ 79%] Building CXX object src/CMakeFiles/exiv2lib.dir/properties.cpp.obj
[ 80%] Building CXX object src/CMakeFiles/exiv2lib.dir/psdimage.cpp.obj
[ 81%] Building CXX object src/CMakeFiles/exiv2lib.dir/quicktimevideo.cpp.obj
[ 83%] Building CXX object src/CMakeFiles/exiv2lib.dir/rafimage.cpp.obj
[ 84%] Building CXX object src/CMakeFiles/exiv2lib.dir/riffvideo.cpp.obj
[ 85%] Building CXX object src/CMakeFiles/exiv2lib.dir/rw2image.cpp.obj
[ 86%] Building CXX object src/CMakeFiles/exiv2lib.dir/tags.cpp.obj
[ 87%] Building CXX object src/CMakeFiles/exiv2lib.dir/tgaimage.cpp.obj
[ 89%] Building CXX object src/CMakeFiles/exiv2lib.dir/tiffimage.cpp.obj
[ 90%] Building CXX object src/CMakeFiles/exiv2lib.dir/types.cpp.obj
[ 91%] Building CXX object src/CMakeFiles/exiv2lib.dir/utilsvideo.cpp.obj
[ 92%] Building CXX object src/CMakeFiles/exiv2lib.dir/value.cpp.obj
[ 93%] Building CXX object src/CMakeFiles/exiv2lib.dir/version.cpp.obj
[ 95%] Building CXX object src/CMakeFiles/exiv2lib.dir/webpimage.cpp.obj
[ 96%] Building CXX object src/CMakeFiles/exiv2lib.dir/xmp.cpp.obj
[ 97%] Building CXX object src/CMakeFiles/exiv2lib.dir/xmpsidecar.cpp.obj
[ 98%] Building CXX object src/CMakeFiles/exiv2lib.dir/pngimage.cpp.obj
[100%] Linking CXX shared library ../bin/libexiv2.dll
[100%] Built target exiv2lib
[gilles@localhost build (master)]$

My patch applied is very simple : 

diff --git a/include/exiv2/config.h b/include/exiv2/config.h
index 41082d83..7b28f15c 100644
--- a/include/exiv2/config.h
+++ b/include/exiv2/config.h
@@ -67,10 +67,10 @@ typedef int pid_t;
 # ifndef  __MINGW__
 #  define __MINGW__ 1
 # endif
-
+/*
 #ifdef EXV_UNICODE_PATH
 #error EXV_UNICODE_PATH is not supported for MinGW builds
-#endif
+#endif*/
 #endif
 
 #ifndef __CYGWIN__

Gilles Caulier

**Comment by clanmills** [comment]
Gilles

What's the issue here?  EXV_UNICODE_PATH didn't work with MinGW msys/1.  However, one of the darktable engineers (Tobias, I think) told me that it works OK with MinGW msys/2.

Are you asking us to remove this MinGW "blocker"?

**Comment by cgilles** [comment]
Absolutly.

i compiled more Exiv2 code, as the CLI tool and all link fine.

This evening, i build the first digiKam 6.0.0 bundle for windows, cross
compiled with MinGW with Exiv2 0.27. I will see if all work fine under
Windows 10 quickly.

Gilles Caulier

Le mar. 9 oct. 2018 à 18:36, Robin Mills <notifications@github.com> a
écrit :

**Comment by clanmills** [comment]
I've removed the MinGW "blocker" in include/exiv2/config.h in branch "RC1"  1f1bb97c589d59991fa23b2cd6eb72ff348ccd2f

This triggered a consequential change to samples/exifprint.cpp which is the only sample which respects EXV_UNICODE_PATH and can pass pathnames in UNICODE.  This involves hackery concerning main() and wmain() and results in linking errors with WinMain()!  

I found this discussion suggesting linking with libmingw32 as follows:  https://github.com/libSDL2pp/libSDL2pp/issues/66
```
$ cmake .. -G "Unix Makefiles" -DEXIV2_ENABLE_WIN_UNICODE=On -DCMAKE_EXE_LINKER_FLAGS="-lmingw32" 
```
That moved the linking issue from samples/exifprint.cpp to src/exiv2.cpp   I took the easy way out and "neutered" samples/exifprint.cpp UNICODE_PATH support.

**Do not use the option `-DCMAKE_EXE_LINKER_FLAGS="-lmingw32" `** 

Build passes the test suite.

**Comment by cgilles** [comment]
Note : I can confirm that Exiv2 0.26 do not link at end when UNICODE option
is enabled with my patch.
In same condition, Exiv2 0.27 compile and link fine. So the new version of
code is compatible, older one no.

Gilles Caulier

Le mar. 9 oct. 2018 à 19:51, Robin Mills <notifications@github.com> a
écrit :

**Comment by clanmills** [comment]
That's correct, Gilles.  exiv2 v0.26 MinGW/msys-1.0 does not support UNICODE_PATH.   Team Exiv2 has never provided support for MinGW/msys2 for Exiv2 v0.26 and earlier.

For v0.27, we will support MinGW/msys2 **and** UNICODE_PATH.  I closed this issue because the fix has been submitted to branch "RC1" 1f1bb97
